File: samples/bot-request-approval/python/server/bots/teamsBot.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License.

import os
from botbuilder.core import TurnContext, CardFactory
from botbuilder.schema import ActivityTypes, Activity
from botbuilder.core.teams import TeamsInfo, TeamsActivityHandler
from server.models import (
    adaptive_card
)

class TeamsBot(TeamsActivityHandler):
    def __init__(self):
        super().__init__()
        self.base_url = os.getenv("BaseUrl")  # Get base URL from environment variables

    async def on_members_added_activity(self, members_added, turn_context: TurnContext):
        # Greet new users when they are added to the conversation
        for member in members_added:
            if member.id != turn_context.activity.recipient.id:
                await turn_context.send_activity(
                    "Hello and welcome! With this sample you can send task requests to your manager, and your manager can approve/reject the request."
                )

    async def on_message_activity(self, turn_context: TurnContext):
        channel_data = turn_context.activity.channel_data  # Channel metadata as dict

        # Suppress Teams schema warnings by accessing known attributes
        source = channel_data.get("source", None)  # Accessing 'source' to avoid warning
        legacy = channel_data.get("legacy", None)  # Accessing 'legacy' to avoid warning

        await self.start_task_management(turn_context)  # Show options card to user

    async def on_invoke_activity(self, turn_context: TurnContext):
        activity = turn_context.activity
        user = activity.from_property  # Current user object

        # Handle Adaptive Card submit actions
        if activity.name == "adaptiveCard/action":
            all_members = await TeamsInfo.get_members(turn_context)  # Get team members
            card = await adaptive_card.select_response_card(
                turn_context, user, all_members  # Build response card based on user
            )
            return adaptive_card.invoke_response(card)  # Return card as response

    async def start_task_management(self, turn_context: TurnContext):
        card = adaptive_card.option_inc()  # Load initial option card
        await turn_context.send_activity(
            Activity(
                type=ActivityTypes.message,
                attachments=[CardFactory.adaptive_card(card)],  # Send Adaptive Card
            )
        )

File: samples/TeamsSDK/Archived/bot-request-approval/nodejs/bots/teamsBot.ts
import { TeamsActivityHandler, CardFactory, TeamsInfo, TurnContext } from 'botbuilder';
import * as adaptiveCards from '../models/adaptiveCard';

export class TeamsBot extends TeamsActivityHandler {
    constructor() {
        super();

        this.onMembersAdded(this.handleMembersAdded.bind(this));
        this.onMessage(this.handleMessage.bind(this));
    }

    async handleMembersAdded(context: TurnContext, next: () => Promise<void>) {
        const membersAdded = context.activity.membersAdded;
        for (const member of membersAdded) {
            if (member.id !== context.activity.recipient.id) {
                await context.sendActivity("Hello and welcome! With this sample you can send task requests to your manager, and your manager can approve/reject the request.");
            }
        }
        await next();
    }

    async handleMessage(context: TurnContext, next: () => Promise<void>) {
        await context.sendActivity({
            attachments: [CardFactory.adaptiveCard(adaptiveCards.optionInc())]
        });
        await next();
    }

    async onInvokeActivity(context: TurnContext) {
        const user = context.activity.from;
        if (context.activity.name === 'adaptiveCard/action') {
            const allMembers = await TeamsInfo.getMembers(context);
            const card = await adaptiveCards.selectResponseCard(context, user, allMembers);
            return adaptiveCards.invokeResponse(card);
        }
    }
}